File: python/21-30/Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head == None or head.next ==None:
            return head
        su = ListNode(0)
        su.next = head
        p = head
        pre,slow,fast =su,p,p.next
        while fast!=None:
            pre.next = fast
            slow.next=fast.next
            fast.next = slow
            #update
            pre=slow
            if slow.next == None or slow.next.next==None:
                break
            else:
                fast =pre.next.next
                slow = pre.next
        return su.next

File: Cpp/21-30/Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 /*
  runtime :4ms
  error: 1
  time complexity : o(n);
 
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* thead = new ListNode(0);
        ListNode* temp = thead;
        while(head!=NULL && head->next!=NULL){
          ListNode* t = head->next->next;
          head->next->next = head;
          temp->next = head->next;
          head->next = NULL;
          temp = temp->next->next;
          head = t;
        }
        return thead->next;
    }
};
